src/frequenz/channels/_receiver.py

```diff
@@ -267,11 +270,67 @@ def map(
         """
         return _Mapper(receiver=self, mapping_function=mapping_function)
 
+    @overload
+    def filter(
+        self,
+        filter_function: Callable[
+            [ReceiverMessageT_co], TypeGuard[FilteredMessageT_co]
+        ],
+        /,
+    ) -> Receiver[FilteredMessageT_co]:
+        """Apply a type guard on the messages on a receiver.
+
+        Tip:
+            The returned receiver type won't have all the methods of the original
+            receiver. If you need to access methods of the original receiver that are
+            not part of the `Receiver` interface you should save a reference to the
+            original receiver and use that instead.
+
+        Args:
+            filter_function: The function to be applied on incoming messages to
+                determine if they should be received.
+
+        Returns:
+            A new receiver that only receives messages that pass the filter.
+        """
+        ...  # pylint: disable=unnecessary-ellipsis
+
+    @overload
     def filter(
         self, filter_function: Callable[[ReceiverMessageT_co], bool], /
     ) -> Receiver[ReceiverMessageT_co]:
         """Apply a filter function on the messages on a receiver.
 
+        Tip:
+            The returned receiver type won't have all the methods of the original
+            receiver. If you need to access methods of the original receiver that are
+            not part of the `Receiver` interface you should save a reference to the
+            original receiver and use that instead.
+
+        Args:
+            filter_function: The function to be applied on incoming messages to
+                determine if they should be received.
+
+        Returns:
+            A new receiver that only receives messages that pass the filter.
+        """
+        ...  # pylint: disable=unnecessary-ellipsis
+
+    # We need to use Any here because otherwise _Filter would have to deal with two
+    # different signatures. We can create two filter classes, one for regular functions
+    # and one for type guards, but then there is no way to tell at runtime which
+    # function is a type guard and which isn't to instantiate the correct class.
+    # Using Any here has no impact though, as thanks to the overloads, only the
+    # overloaded types will be accepted.
+    def filter(
+        self, filter_function: Callable[[ReceiverMessageT_co], Any], /
+    ) -> Receiver[Any]:
```

Comment:
You don't need to change `_Filter` because `TypeGuard` is just another name for bool, with some additional semantic meaning.  The type system knows assigning a `TypeGuard` value to a `bool` is fine.

```suggestion
    def filter(
        self,
        filter_function: (
            Callable[[ReceiverMessageT_co], bool]
            | Callable[[ReceiverMessageT_co], TypeGuard[FilteredMessageT_co]]
        ),
        /,
    ) -> Receiver[ReceiverMessageT_co] | Receiver[FilteredMessageT_co]:
```

Reply:
Interesting, I think what I tried was: `filter_function: Callable[[ReceiverMessageT_co], bool | TypeGuard[FilteredMessageT_co]]` and that didn't work.